### PR TITLE
3dsmax deploy m

### PR DESCRIPTION
**Pull request by TimoBuskeKdab**
* UI for Kuesa Layers plugin
* Extended Kuesa Layers command to increase performance
* Switched off debug messages (performance)
* Python launcher / deployment

### CODE REVIEW

File: KuesaLayers/KuesaLayersCommand.cpp [hunk 1]
```diff
@@ -49,32 +49,37 @@ Value *KuesaLayersCommand::doIt()
...
     layerName = getArgs()->getStringByKey(L"addLayer", foundKey);
+    // Adds specified layer to input nodes.
     if (foundKey) {
         forAllInputNodes([this, &layerName](INode *node) {
             addLayerToNode(node, layerName);
             return true;
         });
     }
     if (getArgs()->getStringArrayByKey(L"addLayers", layerNames)) {
+    // Adds specified layers to input nodes.
```
Comment (mkrus) [inline KuesaLayers/KuesaLayersCommand.cpp]: move outside if {} or indent properly (same below)

File: KuesaLayers/KuesaLayersCommand.cpp [hunk 2]
```diff
@@ -40,7 +40,7 @@
 #include "KdabLog.h"
 
 KuesaLayersCommand::KuesaLayersCommand(Interface *maxInterface, Value **args, int numArgs)
-    : KdabMaxScriptCommand(maxInterface, args, numArgs)
+  : KdabMaxScriptCommand(maxInterface, args, numArgs)
```
Comment (mkrus) [inline KuesaLayers/KuesaLayersCommand.cpp]: why?

File: KuesaLayers/KuesaLayersCommand.cpp [hunk 3]
```diff
@@ -102,6 +108,93 @@ Value *KuesaLayersCommand::doIt()
...
+
+    layerName = getArgs()->getStringByKey(L"replaceLayer", foundKey);
+    // Replaces layer specified by 'replaceLayer' of input nodes by layer specified by 'replaceBy'.
+    if (foundKey) {
+        std::wstring newLayerName = getArgs()->getStringByKey(L"replaceBy", foundKey);
+        if (foundKey) {
+            forAllInputNodes([this, &layerName, &newLayerName](INode *node) {
+                std::set<std::wstring> gatheredLayerNames;
+                gatherLayersFromNodeAppend(node, gatheredLayerNames);
+                if (gatheredLayerNames.find(layerName) != gatheredLayerNames.end() && gatheredLayerNames.find(newLayerName) == gatheredLayerNames.end())
```
Comment (mkrus) [inline KuesaLayers/KuesaLayersCommand.cpp]: long line, split after &&

File: KuesaLayers/KuesaLayersCommand.cpp (hunk 2 above)
Comment (TimoBuskeKdab) [reply KuesaLayers/KuesaLayersCommand.cpp]: oops.

File: KuesaLayers/KuesaLayersCommand.cpp (hunk 1 above)
Comment (TimoBuskeKdab) [reply KuesaLayers/KuesaLayersCommand.cpp]: done

File: KuesaLayers/KuesaLayersCommand.cpp (hunk 3 above)
Comment (TimoBuskeKdab) [reply KuesaLayers/KuesaLayersCommand.cpp]: done

**Comment by TimoBuskeKdab** [comment]
about the dlu... don't know. i thought this was the easiest for the user. but we could also change the solution to copy it there. then the user would have to build...

**Comment by mkrus** [comment]
but we don't ship the exporter dll do we?

Maybe we could, then update the instructions?

**Comment by TimoBuskeKdab** [comment]
i removed the .dlu and set the build path to deploy folder.

**Comment by TimoBuskeKdab** [comment]
i think you are right. it's strange to to the exporter differently

File: README.md [hunk 4]
```diff
@@ -29,3 +29,16 @@ Please set the following environment variables, adjusting paths to match your in
 * Open `KuesaPlugins.sln`
 * If prompted, make sure is used `v140` build tools and `10.0.10586.0` SDK
 
+# Installing Kuesa Plugins For 3DS Max
```
Comment (mkrus) [inline README.md]: these instructions are not quite right I think.

You need to build both projects (which will install the exporter dll in the right place and build the dlu).

Then install the extra deployment plugins.

**Pull request merged**